core/shared/src/main/scala/scalaz/zio/ZSchedule.scala

```diff
@@ -584,6 +584,13 @@ trait Schedule_Functions extends Serializable {
   final def doUntil[A](f: A => Boolean): Schedule[A, A] =
     identity[A].untilInput(f)
 
+  /**
+    * A schedule that recurs for until the input value becomes applicable to partial function
+    * and then map that value with given function.
+    * */
+  final def doUntil[A, B](pf: PartialFunction[A, B]): Schedule[A, B] =
+    doUntil[A](pf.isDefinedAt).map(pf)
```

Comment:
Would it be possible to implement this using `pf.applyOrElse` to avoid double evaluation of `pf`?

Reply:
Thanks, good catch! I'll give it a try.

Reply:
As far as I can see `pf.applyOrElse` requires total function as a fallback and that fallback function domain and codomain should correspond to A and B types hence we could not use `applyOrElse` to fallback to smth like `Schedule[A, B]`. @iravid please could you give an example of how you see this impl with  `pf.applyOrElse` ?

Reply:
Ah, well, this exposes the real issue: `doUntil` with a partial function should yield a `Schedule[A, Option[B]]`.

Reply:
Sorry, that was inaccurate. You could still implement this, but you need to modify the partial function passed to you to return a `Decision`. You could do this with `reconsider`.

Reply:
Sorry, I still can't understand, how can I turn a PF to a total function with `reconsider`. If partial function is not defined on input value it can't return a `Decision` that would contain output of type B, because it didn't get that value from the original function, I can fallback to previous value of type A in a new   `Decision`, but `reconsider` restricts us to a total function that always returns a `Decision` of exactly one output type, either A or B.

Reply:
if you set `C = B` in `reconsider`, you'll see it gives you an `A` and a `Decision[State, B]`. You then need to return a `Decision[State, B]`. You can then compose your partial function with a `Decision` constructor, and apply it or fallback to the `Decision[State, B]` given to you.

Reply:
I think it's impossible to write this safely because `Schedule`'s contract is that for each `A`, it can produce a `B`. Now it so happens the `B` is evaluated lazily. And it also happens that `repeat` won't pull out a `B` until the schedule is done. So it would be safe for `repeat`, but not for `retry`. And I think `repeatOrElse` will pull out the `B` in an event of an error.

Overall I think we cannot do this API as spec'd out safely. The best we can do is:

```scala
doUntil[A](pf.isDefinedAt).map(pf.lift)
```

which of course can be done in `reconsider` to avoid evaluating `pf` twice.

This yields an `Option[B]`, which the user can convert to a `B` inside of a suitable `ZIO`.